File: ios/MapTalk/Data/MediaUploader.swift
import FirebaseAuth
import FirebaseStorage
import Foundation

/// Uploads chat media and returns the public URL stored on the message document.
protocol MediaUploading: AnyObject {
    func upload(threadId: String, messageId: String, image: PreparedImage) async throws -> String
    func upload(threadId: String, messageId: String, audio: PreparedAudio) async throws -> String
    func upload(threadId: String, messageId: String, video: PreparedVideo) async throws -> String
}

final class R2MediaUploader: MediaUploading {
    private let auth: Auth
    private let imageEndpoint: URL
    private let audioEndpoint: URL
    private let videoEndpoint: URL
    private let uploadGate = UploadGate(limit: 2)

    init(auth: Auth, endpoint: URL) {
        self.auth = auth
        self.imageEndpoint = endpoint
        // …/v1/images → …/v1/audio and …/v1/video
        var audio = endpoint
        var video = endpoint
        if endpoint.lastPathComponent == "images" {
            audio.deleteLastPathComponent()
            audio.appendPathComponent("audio")
            video.deleteLastPathComponent()
            video.appendPathComponent("video")
        } else {
            audio = URL(string: "https://maptalk-media.hhypkfpshg.workers.dev/v1/audio")!
            video = URL(string: "https://maptalk-media.hhypkfpshg.workers.dev/v1/video")!
        }
        self.audioEndpoint = audio
        self.videoEndpoint = video
    }

    func upload(threadId: String, messageId: String, image: PreparedImage) async throws -> String {
        await uploadGate.acquire()
        do {
            let url = try await post(
                endpoint: imageEndpoint,
                threadId: threadId,
                messageId: messageId,
                body: image.jpegData,
                contentType: "image/jpeg"
            )
            await uploadGate.release()
            return url
        } catch {
            await uploadGate.release()
            throw error
        }
    }

    func upload(threadId: String, messageId: String, audio: PreparedAudio) async throws -> String {
        await uploadGate.acquire()
        do {
            let url = try await post(
                endpoint: audioEndpoint,
                threadId: threadId,
                messageId: messageId,
                body: audio.data,
                contentType: audio.contentType
            )
            await uploadGate.release()
            return url
        } catch {
            await uploadGate.release()
            throw error
        }
    }

    func upload(threadId: String, messageId: String, video: PreparedVideo) async throws -> String {
        await uploadGate.acquire()
        do {
            let url = try await putFile(
                endpoint: videoEndpoint,
                threadId: threadId,
                messageId: messageId,
                fileURL: video.fileURL,
                contentType: video.contentType,
                extraHeaders: ["X-MapTalk-Duration-Ms": "\(video.durationMs)"]
            )
            await uploadGate.release()
            return url
        } catch {
            await uploadGate.release()
            throw error
        }
    }

    private func post(
        endpoint: URL,
        threadId: String,
        messageId: String,
        body: Data,
        contentType: String,
        extraHeaders: [String: String] = [:],
        maxAttempts: Int = 3
    ) async throws -> String {
        var lastError: Error?
        for attempt in 0..<maxAttempts {
            try Task.checkCancellation()
            do {
                return try await postOnce(
                    endpoint: endpoint,
                    threadId: threadId,
                    messageId: messageId,
                    body: body,
                    contentType: contentType,
                    extraHeaders: extraHeaders
                )
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                lastError = error
                guard Self.isTransient(error), attempt < maxAttempts - 1 else { throw error }
                try await Task.sleep(nanoseconds: UInt64(400_000_000) << attempt)
            }
        }
        throw lastError ?? URLError(.unknown)
    }

    private func putFile(
        endpoint: URL,
        threadId: String,
        messageId: String,
        fileURL: URL,
        contentType: String,
        extraHeaders: [String: String] = [:],
        maxAttempts: Int = 3
    ) async throws -> String {
        var lastError: Error?
        for attempt in 0..<maxAttempts {
            try Task.checkCancellation()
            do {
                return try await putFileOnce(
                    endpoint: endpoint,
                    threadId: threadId,
                    messageId: messageId,
                    fileURL: fileURL,
                    contentType: contentType,
                    extraHeaders: extraHeaders
                )
            } catch is CancellationError {
                throw CancellationError()
            } catch {
                lastError = error
                guard Self.isTransient(error), attempt < maxAttempts - 1 else { throw error }
                try await Task.sleep(nanoseconds: UInt64(400_000_000) << attempt)
            }
        }
        throw lastError ?? URLError(.unknown)
    }

    private func postOnce(
        endpoint: URL,
        threadId: String,
        messageId: String,
        body: Data,
        contentType: String,
        extraHeaders: [String: String]
    ) async throws -> String {
        guard let user = auth.currentUser else {
            throw URLError(.userAuthenticationRequired)
        }
        let token = try await user.getIDToken()
        var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "threadId", value: threadId),
            URLQueryItem(name: "messageId", value: messageId),
        ]
        var request = URLRequest(url: components.url!)
        request.httpMethod = "POST"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        for (key, value) in extraHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }
        request.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: request)
        return try Self.parseUploadURL(data: data, response: response)
    }

    private func putFileOnce(
        endpoint: URL,
        threadId: String,
        messageId: String,
        fileURL: URL,
        contentType: String,
        extraHeaders: [String: String]
    ) async throws -> String {
        guard let user = auth.currentUser else {
            throw URLError(.userAuthenticationRequired)
        }
        let token = try await user.getIDToken()
        var components = URLComponents(url: endpoint, resolvingAgainstBaseURL: false)!
        components.queryItems = [
            URLQueryItem(name: "threadId", value: threadId),
            URLQueryItem(name: "messageId", value: messageId),
        ]
        var request = URLRequest(url: components.url!)
        request.httpMethod = "PUT"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        let length = (try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.size] as? NSNumber)?.intValue ?? 0
        request.setValue("\(length)", forHTTPHeaderField: "Content-Length")
        for (key, value) in extraHeaders {
            request.setValue(value, forHTTPHeaderField: key)
        }

        let (data, response) = try await URLSession.shared.upload(for: request, fromFile: fileURL)
        return try Self.parseUploadURL(data: data, response: response)
    }

    private static func parseUploadURL(data: Data, response: URLResponse) throws -> String {
        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        guard (200...299).contains(http.statusCode) else {
            let detail = String(data: data, encoding: .utf8) ?? ""
            throw NSError(
                domain: "MapTalk.MediaUploader",
                code: http.statusCode,
                userInfo: [NSLocalizedDescriptionKey: "Upload failed (\(http.statusCode)): \(detail)"]
            )
        }
        return try JSONDecoder().decode(UploadResponse.self, from: data).url
    }

    private static func isTransient(_ error: Error) -> Bool {
        if let urlError = error as? URLError {
            switch urlError.code {
            case .timedOut, .networkConnectionLost, .notConnectedToInternet, .cannotConnectToHost:
                return true
            default:
                break
            }
        }
        let ns = error as NSError
        return (500...599).contains(ns.code)
    }

    private struct UploadResponse: Decodable {
        let url: String
    }
}

/// Limits parallel R2 uploads so a busy thread doesn't stampede the Worker.
private actor UploadGate {
    private let limit: Int
    private var inFlight = 0
    private var waiters: [CheckedContinuation<Void, Never>] = []

    init(limit: Int) {
        self.limit = limit
    }

    func acquire() async {
        while inFlight >= limit {
            await withCheckedContinuation { cont in
                waiters.append(cont)
            }
        }
        inFlight += 1
    }

    func release() {
        inFlight = max(0, inFlight - 1)
        guard !waiters.isEmpty else { return }
        waiters.removeFirst().resume()
    }
}

final class FirebaseStorageUploader: MediaUploading {
    private let storage: Storage

    init(storage: Storage) {
        self.storage = storage
    }

    func upload(threadId: String, messageId: String, image: PreparedImage) async throws -> String {
        let ref = storage.reference()
            .child("threads")
            .child(threadId)
            .child("\(messageId).jpg")
        let metadata = StorageMetadata()
        metadata.contentType = "image/jpeg"
        _ = try await ref.putDataAsync(image.jpegData, metadata: metadata)
        return try await ref.downloadURL().absoluteString
    }

    func upload(threadId: String, messageId: String, audio: PreparedAudio) async throws -> String {
        let ref = storage.reference()
            .child("threads")
            .child(threadId)
            .child("\(messageId).m4a")
        let metadata = StorageMetadata()
        metadata.contentType = audio.contentType
        _ = try await ref.putDataAsync(audio.data, metadata: metadata)
        return try await ref.downloadURL().absoluteString
    }

    func upload(threadId: String, messageId: String, video: PreparedVideo) async throws -> String {
        let ref = storage.reference()
            .child("threads")
            .child(threadId)
            .child("\(messageId).mp4")
        let metadata = StorageMetadata()
        metadata.contentType = video.contentType
        _ = try await ref.putFileAsync(from: video.fileURL, metadata: metadata)
        return try await ref.downloadURL().absoluteString
    }
}

File: android/app/src/main/java/app/maptalk/data/MediaUploader.kt
package app.maptalk.data

import app.maptalk.data.model.PreparedAudio
import app.maptalk.data.model.PreparedImage
import app.maptalk.data.model.PreparedVideo
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageMetadata
import java.io.IOException
import java.net.HttpURLConnection
import java.net.SocketTimeoutException
import java.net.URL
import java.net.URLEncoder
import java.net.UnknownHostException
import kotlin.coroutines.cancellation.CancellationException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.sync.Semaphore
import kotlinx.coroutines.sync.withPermit
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext
import org.json.JSONObject

/**
 * Uploads chat media and returns the public URL stored on the message document.
 *
 * Live builds POST to the Cloudflare Worker (R2, zero egress). Emulator builds use Firebase
 * Storage so media still works against `scripts/mock-data.sh` without Cloudflare.
 */
sealed class MediaUploader {
    abstract suspend fun upload(threadId: String, messageId: String, image: PreparedImage): String
    abstract suspend fun upload(threadId: String, messageId: String, audio: PreparedAudio): String
    abstract suspend fun upload(threadId: String, messageId: String, video: PreparedVideo): String

    class R2(
        private val auth: FirebaseAuth,
        private val imageEndpoint: String,
    ) : MediaUploader() {
        private val uploadSlots = Semaphore(MAX_CONCURRENT_UPLOADS)

        private val audioEndpoint: String =
            if (imageEndpoint.endsWith("/v1/images")) {
                imageEndpoint.removeSuffix("/v1/images") + "/v1/audio"
            } else {
                "https://maptalk-media.hhypkfpshg.workers.dev/v1/audio"
            }

        private val videoEndpoint: String =
            if (imageEndpoint.endsWith("/v1/images")) {
                imageEndpoint.removeSuffix("/v1/images") + "/v1/video"
            } else {
                "https://maptalk-media.hhypkfpshg.workers.dev/v1/video"
            }

        private val videoPresignEndpoint: String = "$videoEndpoint/presign"
        private val videoConfirmEndpoint: String = "$videoEndpoint/confirm"

        override suspend fun upload(threadId: String, messageId: String, image: PreparedImage): String =
            uploadSlots.withPermit {
                post(imageEndpoint, threadId, messageId, image.jpegBytes, "image/jpeg")
            }

        override suspend fun upload(threadId: String, messageId: String, audio: PreparedAudio): String =
            uploadSlots.withPermit {
                post(audioEndpoint, threadId, messageId, audio.bytes, audio.contentType)
            }

        override suspend fun upload(threadId: String, messageId: String, video: PreparedVideo): String =
            uploadSlots.withPermit {
                uploadVideo(threadId, messageId, video)
            }

        private suspend fun uploadVideo(
            threadId: String,
            messageId: String,
            video: PreparedVideo,
        ): String = withContext(Dispatchers.IO) {
            val durationHeader = mapOf("X-MapTalk-Duration-Ms" to video.durationMs.toString())
            val length = video.file.length()
            if (length <= 0L) throw IOException("Video file is empty")

            val presign = tryPresign(threadId, messageId, length, durationHeader)
            if (presign != null) {
                putToPresignedUrl(presign.uploadUrl, video.file, presign.contentType)
                return@withContext postConfirm(threadId, messageId, durationHeader)
            }

            // Worker without R2 S3 secrets (or old deploy): stream through Worker.
            putFile(
                videoEndpoint,
                threadId,
                messageId,
                video.file,
                video.contentType,
                extraHeaders = durationHeader,
            )
        }

        private suspend fun tryPresign(
            threadId: String,
            messageId: String,
            contentLength: Long,
            extraHeaders: Map<String, String>,
        ): PresignResponse? {
            val token = auth.currentUser?.getIdToken(false)?.await()?.token
                ?: throw IllegalStateException("Sign in before sending media")
            val url = URL(
                "$videoPresignEndpoint?threadId=${enc(threadId)}&messageId=${enc(messageId)}",
            )
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = false
                connectTimeout = 30_000
                readTimeout = 30_000
                setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("X-MapTalk-Content-Length", contentLength.toString())
                extraHeaders.forEach { (k, v) -> setRequestProperty(k, v) }
            }
            try {
                val code = connection.responseCode
                val body = (if (code in 200..299) connection.inputStream else connection.errorStream)
                    ?.bufferedReader()
                    ?.use { it.readText() }
                    .orEmpty()
                if (code == 503 && body.contains("presign_unavailable")) {
                    return null
                }
                if (code !in 200..299) {
                    throw IOException("Media upload failed ($code): $body")
                }
                val json = JSONObject(body)
                return PresignResponse(
                    uploadUrl = json.getString("uploadUrl"),
                    contentType = json.optString("contentType", "video/mp4"),
                )
            } finally {
                connection.disconnect()
            }
        }

        private fun putToPresignedUrl(uploadUrl: String, file: java.io.File, contentType: String) {
            val length = file.length()
            val connection = (URL(uploadUrl).openConnection() as HttpURLConnection).apply {
                requestMethod = "PUT"
                doOutput = true
                connectTimeout = 30_000
                readTimeout = 120_000
                setFixedLengthStreamingMode(length)
                setRequestProperty("Content-Type", contentType)
                setRequestProperty("Content-Length", length.toString())
            }
            try {
                connection.outputStream.use { out ->
                    file.inputStream().use { input ->
                        input.copyTo(out, bufferSize = 64 * 1024)
                    }
                }
                val code = connection.responseCode
                if (code !in 200..299) {
                    val err = connection.errorStream?.bufferedReader()?.use { it.readText() }.orEmpty()
                    throw IOException("Media upload failed ($code): $err")
                }
            } finally {
                connection.disconnect()
            }
        }

        private suspend fun postConfirm(
            threadId: String,
            messageId: String,
            extraHeaders: Map<String, String>,
            maxAttempts: Int = 3,
        ): String {
            var lastError: Exception? = null
            repeat(maxAttempts) { attempt ->
                try {
                    return postEmpty(videoConfirmEndpoint, threadId, messageId, extraHeaders)
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    lastError = e
                    if (!isTransientUploadError(e) || attempt == maxAttempts - 1) throw e
                    delay(400L * (1L shl attempt))
                }
            }
            throw lastError ?: IOException("Media upload failed")
        }

        private suspend fun postEmpty(
            endpoint: String,
            threadId: String,
            messageId: String,
            extraHeaders: Map<String, String>,
        ): String {
            val token = auth.currentUser?.getIdToken(false)?.await()?.token
                ?: throw IllegalStateException("Sign in before sending media")
            val url = URL(
                "$endpoint?threadId=${enc(threadId)}&messageId=${enc(messageId)}",
            )
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = false
                connectTimeout = 30_000
                readTimeout = 60_000
                setRequestProperty("Authorization", "Bearer $token")
                extraHeaders.forEach { (k, v) -> setRequestProperty(k, v) }
            }
            try {
                return readUploadUrl(connection)
            } finally {
                connection.disconnect()
            }
        }

        private data class PresignResponse(
            val uploadUrl: String,
            val contentType: String,
        )

        private suspend fun post(
            endpoint: String,
            threadId: String,
            messageId: String,
            body: ByteArray,
            contentType: String,
            extraHeaders: Map<String, String> = emptyMap(),
            maxAttempts: Int = 3,
        ): String = withContext(Dispatchers.IO) {
            var lastError: Exception? = null
            repeat(maxAttempts) { attempt ->
                ensureActive()
                try {
                    return@withContext postOnce(
                        endpoint,
                        threadId,
                        messageId,
                        body,
                        contentType,
                        extraHeaders,
                    )
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    lastError = e
                    if (!isTransientUploadError(e) || attempt == maxAttempts - 1) throw e
                    delay(400L * (1L shl attempt))
                }
            }
            throw lastError ?: IOException("Media upload failed")
        }

        private suspend fun putFile(
            endpoint: String,
            threadId: String,
            messageId: String,
            file: java.io.File,
            contentType: String,
            extraHeaders: Map<String, String> = emptyMap(),
            maxAttempts: Int = 3,
        ): String = withContext(Dispatchers.IO) {
            var lastError: Exception? = null
            repeat(maxAttempts) { attempt ->
                ensureActive()
                try {
                    return@withContext putFileOnce(
                        endpoint,
                        threadId,
                        messageId,
                        file,
                        contentType,
                        extraHeaders,
                    )
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    lastError = e
                    if (!isTransientUploadError(e) || attempt == maxAttempts - 1) throw e
                    delay(400L * (1L shl attempt))
                }
            }
            throw lastError ?: IOException("Media upload failed")
        }

        private suspend fun postOnce(
            endpoint: String,
            threadId: String,
            messageId: String,
            body: ByteArray,
            contentType: String,
            extraHeaders: Map<String, String>,
        ): String {
            val token = auth.currentUser?.getIdToken(false)?.await()?.token
                ?: throw IllegalStateException("Sign in before sending media")
            val url = URL(
                "$endpoint?threadId=${enc(threadId)}&messageId=${enc(messageId)}",
            )
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = true
                connectTimeout = 30_000
                readTimeout = 60_000
                setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("Content-Type", contentType)
                setRequestProperty("Content-Length", body.size.toString())
                extraHeaders.forEach { (k, v) -> setRequestProperty(k, v) }
            }
            try {
                connection.outputStream.use { it.write(body) }
                return readUploadUrl(connection)
            } finally {
                connection.disconnect()
            }
        }

        private suspend fun putFileOnce(
            endpoint: String,
            threadId: String,
            messageId: String,
            file: java.io.File,
            contentType: String,
            extraHeaders: Map<String, String>,
        ): String {
            val token = auth.currentUser?.getIdToken(false)?.await()?.token
                ?: throw IllegalStateException("Sign in before sending media")
            val url = URL(
                "$endpoint?threadId=${enc(threadId)}&messageId=${enc(messageId)}",
            )
            val length = file.length()
            val connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "PUT"
                doOutput = true
                connectTimeout = 30_000
                readTimeout = 120_000
                setFixedLengthStreamingMode(length)
                setRequestProperty("Authorization", "Bearer $token")
                setRequestProperty("Content-Type", contentType)
                setRequestProperty("Content-Length", length.toString())
                extraHeaders.forEach { (k, v) -> setRequestProperty(k, v) }
            }
            try {
                connection.outputStream.use { out ->
                    file.inputStream().use { input ->
                        input.copyTo(out, bufferSize = 64 * 1024)
                    }
                }
                return readUploadUrl(connection)
            } finally {
                connection.disconnect()
            }
        }

        private fun readUploadUrl(connection: HttpURLConnection): String {
            val code = connection.responseCode
            val responseBody = (if (code in 200..299) connection.inputStream else connection.errorStream)
                ?.bufferedReader()
                ?.use { it.readText() }
                .orEmpty()
            if (code !in 200..299) {
                throw IOException("Media upload failed ($code): $responseBody")
            }
            return JSONObject(responseBody).getString("url")
        }

        private fun isTransientUploadError(error: Exception): Boolean {
            val msg = error.message.orEmpty()
            return msg.contains("Media upload failed (5") ||
                msg.contains("timeout", ignoreCase = true) ||
                error is SocketTimeoutException ||
                error is UnknownHostException
        }

        private fun enc(value: String): String =
            URLEncoder.encode(value, Charsets.UTF_8.name())

        companion object {
            private const val MAX_CONCURRENT_UPLOADS = 2
        }
    }

    class Firebase(
        private val storage: FirebaseStorage,
    ) : MediaUploader() {
        override suspend fun upload(threadId: String, messageId: String, image: PreparedImage): String {
            val ref = storage.reference
                .child("threads")
                .child(threadId)
                .child("$messageId.jpg")
            val metadata = StorageMetadata.Builder()
                .setContentType("image/jpeg")
                .build()
            putBytes(ref, image.jpegBytes, metadata)
            return downloadUrl(ref)
        }

        override suspend fun upload(threadId: String, messageId: String, audio: PreparedAudio): String {
            val ref = storage.reference
                .child("threads")
                .child(threadId)
                .child("$messageId.m4a")
            val metadata = StorageMetadata.Builder()
                .setContentType(audio.contentType)
                .build()
            putBytes(ref, audio.bytes, metadata)
            return downloadUrl(ref)
        }

        override suspend fun upload(threadId: String, messageId: String, video: PreparedVideo): String {
            val ref = storage.reference
                .child("threads")
                .child(threadId)
                .child("$messageId.mp4")
            val metadata = StorageMetadata.Builder()
                .setContentType(video.contentType)
                .build()
            putFile(ref, video.file, metadata)
            return downloadUrl(ref)
        }

        private suspend fun putBytes(
            ref: com.google.firebase.storage.StorageReference,
            bytes: ByteArray,
            metadata: StorageMetadata,
        ) {
            suspendCancellableCoroutine { cont ->
                val task = ref.putBytes(bytes, metadata)
                cont.invokeOnCancellation { task.cancel() }
                task
                    .addOnSuccessListener { cont.resume(Unit) }
                    .addOnFailureListener { cont.resumeWithException(it) }
            }
        }

        private suspend fun putFile(
            ref: com.google.firebase.storage.StorageReference,
            file: java.io.File,
            metadata: StorageMetadata,
        ) {
            suspendCancellableCoroutine { cont ->
                val task = ref.putFile(android.net.Uri.fromFile(file), metadata)
                cont.invokeOnCancellation { task.cancel() }
                task
                    .addOnSuccessListener { cont.resume(Unit) }
                    .addOnFailureListener { cont.resumeWithException(it) }
            }
        }

        private suspend fun downloadUrl(ref: com.google.firebase.storage.StorageReference): String =
            suspendCancellableCoroutine { cont ->
                ref.downloadUrl
                    .addOnSuccessListener { uri -> cont.resume(uri.toString()) }
                    .addOnFailureListener { cont.resumeWithException(it) }
            }
    }
}
